Source/WebCore/platform/graphics/BitmapImageSource.cpp

```diff
@@ -0,0 +1,484 @@
+    m_frameProvider->destroyDecodedData(destroyAll);
+}
+
+bool BitmapImageSource::canDestroyNativeImageAtIndex(unsigned index)
+{
+    return !(isAnimated() && index == currentFrameIndex());
+}
+
+void BitmapImageSource::decodedSizeChanged(long long decodedSize)
+{
+    if (!decodedSize)
+        return;
+
+    if (auto observer = m_bitmapImage.imageObserver())
+        observer->decodedSizeChanged(m_bitmapImage, decodedSize);
+}
+
+void BitmapImageSource::resetData()
+{
+    if (auto data = m_bitmapImage.data())
+        m_decoder->setData(*data, m_decoder->isAllDataReceived());
+}
+
+void BitmapImageSource::startAnimation()
+{
+    startAnimation(SubsamplingLevel::Default, DecodingMode::Synchronous);
+}
+
+bool BitmapImageSource::startAnimation(SubsamplingLevel subsamplingLevel, const DecodingOptions& options)
+{
+    if (auto animator = ensureAnimator())
+        return animator->startAnimation(subsamplingLevel, options);
+    return false;
+}
+
+void BitmapImageSource::stopAnimation()
+{
+    if (!m_animator)
+        return;
+
+    m_animator->stopAnimation();
+    stopDecodingWorkQueue();
+}
+
+void BitmapImageSource::resetAnimation()
+{
+    m_animator = nullptr;
+}
+
+bool BitmapImageSource::isAnimated() const
+{
+    return frameCount() > 1 && repetitionCount() != RepetitionCountNone;
+}
+
+bool BitmapImageSource::isAnimating() const
+{
+    return m_animator && m_animator->isAnimating();
+}
+
+bool BitmapImageSource::isAnimationAllowed() const
+{
+    if (m_animator && !m_animator->isAnimationAllowed())
+        return false;
+
+    // ImageObserver may disallow animation.
+    if (auto observer = m_bitmapImage.imageObserver())
+        return observer->allowsAnimation(m_bitmapImage);
+
+    return true;
+}
+
+bool BitmapImageSource::isLargeForDecoding() const
+{
+    size_t sizeInBytes = (const_cast<BitmapImageSource&>(*this)).size(ImageOrientation::Orientation::None).area() * sizeof(uint32_t);
+    return sizeInBytes > (isAnimated() ? 100 * KB : 500 * KB);
+}
+
+void BitmapImageSource::stopDecodingWorkQueue()
+{
+    if (!m_frameProvider)
+        return;
+    m_frameProvider->stopDecodingWorkQueue();
+}
+
+void BitmapImageSource::decode(Function<void(DecodingStatus)>&& decodeCallback)
+{
+    auto frameProvider = ensureFrameProvider();
+    if (!frameProvider)
+        return;
+
+    m_decodeCallbacks.append(WTFMove(decodeCallback));
+    unsigned index = currentFrameIndex();
+
+    if (frameProvider->isPendingDecodingAtIndex(index, SubsamplingLevel::Default, DecodingMode::Asynchronous)) {
+        LOG(Images, "BitmapImageSource::%s - %p - url: %s. Frame at index = %d is being decoded.", __FUNCTION__, this, sourceUTF8(), index);
+        return;
+    }
+
+    bool isCompatibleNativeImage = frameProvider->isCompatibleWithOptionsAtIndex(index, SubsamplingLevel::Default, DecodingMode::Asynchronous);
+    auto animator = ensureAnimator();
+
+    if (animator && (animator->hasEverAnimated() || isCompatibleNativeImage)) {
```

Comment:
I don't quite understand why we check `hasEverAnimated` here. Can you explain this condition that decides whether to call `startAnimation`?

Reply:
The reason for adding `hasEverAnimated` is `startAnimation()` always decodes the `nextFrame` which is `currentFrameIndex + 1`. To need to fall back to the next if-statement the first time the `decode()` method is called for this image. The following frames will be decoded through `startAnimation()`. I added a comment to explain this case.

Reply:
Why do we also have `isCompatibleNativeImage` in the condition? Is it because the `decode()` call in that case means "decode the next frame", and since we already have the current frame decoded, we want to start the animation for the next frame?

Reply:
I think part of my confusion here is what the semantics of `decode()` actually are.

Reply:
If we walk through three consecutive calls to decode():

```
decode()    hasEverAnimated    isCompatibleNativeImage    call
----------------------------------------------------------------------
0           false                  false                  requestNativeImageAtIndex()
1           false                  true                   startAnimation()
2           true                   true                   startAnimation() 
```

And this is what we want. We want frame 0, when decoded  for the first time, to be decoded as soon as possible. But we want for each following frame to be available after the duration of the previous frame.